### ISSUE TITLE
Incorrect function type signature for ParameterDecorator

### ISSUE THREAD

**Issue by bajtos** [microsoft/TypeScript#33260]
**TypeScript Version:**  3.7.0-dev.20190905

**Search Terms:** ParameterDecorator propertyKey

**Code**

The type `ParameterDecorator` is defined as follows in `src/lib/es5.d.ts`:

https://github.com/microsoft/TypeScript/blob/72bb4c2bdcb0cb245c397cd9a8f1f0d1158c6a5d/src/lib/es5.d.ts#L1362

However, at runtime, constructor parameter decorators are invoked with `propertyKey` set to `undefined`.

Consider the following code:

```ts
function inject() {
  return function i(target: Object,
    propertyKey: string | symbol,
    parameterIndex: number
  ): void {
    console.log('propertyKey', typeof propertyKey);
  }
}

class FooBar {
  constructor(
    @inject()
    public name: string
  ) {}
}
```

Compile it with `experimentalDecorators` enabled in `tsconfig.json`.

**Expected behavior:**

Either:
- The code does not compile, because the decorator function must accept `string | symbol | undefined` in the `propertyKey` argument
- Or the program prints `propertyKey string` or `propertyKey symbol` at runtime

**Actual behavior:**
- The code compiles
- The programs prints `undefined` at runtime

**Playground Link:** 

http://www.typescriptlang.org/play/index.html?experimentalDecorators=true#code/GYVwdgxgLglg9mABDMArAptAFASkQbwChFEAndKEUpUSWBZLKAQ1IHMKAuRAeQCMM0ADTESiAA6k449KSgBPANLp53AM5RSKNogA+iNfIC2fOABsRYia2ZGKsgJJgAJugAe3MCBOzRObgBucDDOBKIkEAhq5ugAdGZwbFgA5JLSsgrK8slCiAoycMASUjJySio4ANyiAL6EdYQQZsxqaogAYnBwAEKsYRFRmiDQcKRY4YgAAiiCULgT4iB8ZjAQiGC26Oqa2n4EdXVAA

**Related Issues:** 

- https://github.com/microsoft/TypeScript/pull/2635
- https://github.com/microsoft/TypeScript/issues/20931
- https://github.com/microsoft/TypeScript/issues/8126

**Comment by bajtos** [comment]
I find the current behavior problematic because the type definition for `ParameterDecorator` are not matching the actual runtime behavior.

I'd like to propose two different solutions for your consideration:

1. Change the `ParameterDecorator` type - add `undefined` to allowed `propertyKey` values.

    ```ts
    declare type ParameterDecorator = (target: Object, propertyKey: string | symbol | undefined, parameterIndex: number) => void; 

2. Modify the runtime to set `propertyKey` to an empty string `''` instead of `undefined` when calling parameter decorators for constructor arguments.

Personally, I prefer the first one. I'd rather have my existing codebase to fail the compilation rather than discover a subtle change in TS runtime in production.

**Comment by jakebailey** [comment]
This was fixed back in TS 5.0 and is mentioned in the blog: https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#more-accurate-type-checking-for-parameter-decorators-in-constructors-under-experimentaldecorators
